js/private/js_devserver.js

```diff
@@ -0,0 +1,161 @@
+            return
+        }
+        synced.set(src, lstat.mtimeMs)
+        const exists = fs.existsSync(dst)
+        if (lstat.isSymbolicLink()) {
+            console.log(
+                `Syncing symlink ${src.slice(RUNFILES_ROOT.length + 1)}`
+            )
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.symlink(src, dst)
+        } else if (lstat.isDirectory()) {
+            await mkdirp(dst)
+            const contents = await fs.promises.readdir(src)
+            await Promise.all(
+                contents.map((entry) =>
+                    syncRecursive(path.join(src, entry), path.join(dst, entry))
+                )
+            )
+        } else {
+            console.log(`Syncing file ${src.slice(RUNFILES_ROOT.length + 1)}`)
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.copyFile(src, dst)
+            const mode = isExecutable(lstat) ? 0o744 : 0x644
+            await fs.promises.chmod(dst, mode)
+        }
+    } catch (e) {
+        console.error(e)
+        process.exit(1)
+    }
+}
+
+async function sync(files, rootdir, sync = false) {
+    for (const file of files) {
+        const src = path.join(RUNFILES_ROOT, file)
+        const dst = path.join(rootdir, file)
+        if (sync) {
+            await syncRecursive(src, dst)
+        } else {
+            syncRecursive(src, dst)
+        }
+    }
+}
+
+async function main(args, rootdir) {
+    console.log(`\n\nStarting js_devserver ${process.env.JS_BINARY__TARGET}`)
+
+    const configPath = path.join(RUNFILES_ROOT, args[0])
+
+    const config = JSON.parse(await fs.promises.readFile(configPath))
+
+    await sync(config.data_files, rootdir, true)
+    console.log('Initial sync done')
+
+    return new Promise((resolve) => {
+        const cwd = process.env.JS_BINARY__CHDIR
+            ? path.join(rootdir, process.env.JS_BINARY__CHDIR)
+            : rootdir
+
+        const tool = config.tool
+            ? path.join(RUNFILES_ROOT, config.tool)
+            : config.command
+
+        const toolArgs = args.slice(1)
+
+        console.log(`Running '${tool} ${toolArgs.join(' ')}' in ${cwd}\n\n`)
+
+        const proc = child_process.spawn(tool, toolArgs, {
+            cwd: cwd,
+            stdio: 'inherit',
+            env: {
+                ...process.env,
+                BAZEL_BINDIR: '.',
+                JS_BINARY__CHDIR: '',
+            },
+        })
+
+        proc.on('close', (code) => {
+            console.error(`child tool process exited with code ${code}`)
+            resolve()
+            process.exit(code)
+        })
+
+        process.stdin.on('data', (chunk) => {
+            const chunkString = chunk.toString()
+            if (chunkString.indexOf('IBAZEL_BUILD_COMPLETED SUCCESS') !== -1) {
```

Comment:
What does the string look like such that we can't use `===`?

Could use `chunkString.includes('...')`

Reply:
I don't think there is any guarantee that the chunks in stdin stream won't contain multiple messages from ibazel